src/transformers/models/htdemucs/modeling_htdemucs.py

```diff
@@ -0,0 +1,950 @@
+
+import torch
+import torch.nn as nn
+
+from ...activations import ACT2FN
+from ...modeling_outputs import ModelOutput
+from ...modeling_utils import PreTrainedModel
+from .configuration_htdemucs import HtdemucsConfig
+
+
+@dataclass
+class HtdemucsBaseModelOutput(ModelOutput):
+    """
+    Base class for Htdemuc's model outputs, with potential hidden states and attentions.
+
+    Args:
+        last_freq_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
+            Sequence of frequency hidden-states at the output of the last layer of the model.
+        last_temp_hidden_state (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`):
+            Sequence of temporal hidden-states at the output of the last layer of the model.
+        freq_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
+
+            Hidden-states of the frequency branch of the model at the output of each layer plus the optional initial embedding outputs.
+        temp_hidden_states (`tuple(torch.FloatTensor)`, *optional*, returned when `output_hidden_states=True` is passed or when `config.output_hidden_states=True`):
+            Tuple of `torch.FloatTensor` (one for the output of the embeddings, if the model has an embedding layer, +
+            one for the output of each layer) of shape `(batch_size, sequence_length, hidden_size)`.
+
+            Hidden-states of the frequency branch of the model at the output of each layer plus the optional initial embedding outputs.
+        freq_attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`.
+
+            Attention weights for the frequency branch after the attention softmax, used to compute the weighted average
+            in the self-attention heads.
+        temp_attentions (`tuple(torch.FloatTensor)`, *optional*, returned when `output_attentions=True` is passed or when `config.output_attentions=True`):
+            Tuple of `torch.FloatTensor` (one for each layer) of shape `(batch_size, num_heads, sequence_length,
+            sequence_length)`.
+
+            Attention weights for the frequency branch after the attention softmax, used to compute the weighted average
+            in the self-attention heads.
+    """
+
+    last_freq_hidden_state: torch.FloatTensor = None
+    last_temp_hidden_state: torch.FloatTensor = None
+    freq_hidden_states: Optional[Tuple[torch.FloatTensor]] = None
```

Comment:
This is a bit of a design question for @ylacombe and @ArthurZucker. HTDemucs tracks **two** branches through all the convolutional and Transformer layers: a frequency domain branch (Z), and a temporal domain branch (T). This is different to most audio models, which only use **one** branch (e.g. temporal for Wav2Vec2, or frequency for Whisper). Each branch has its own set of corresponding attention and hidden-states, as shown in the diagram below:

![](https://github.com/facebookresearch/demucs/blob/main/demucs.png)

To me, the most intuitive design for the user is returning **separate** tuples for the frequency/temporal hidden-states, rather than merging them into one tuple. This is because the dimensions are different for each branch, so squashing them into one would give different dimensions for the odd/even indices depending on whether the states are from the frequency/temporal branches.

However, this does violate the Transformers design a bit, since we assume generic outputs of `hidden_states` and `attentions` (not `freq_hidden_states`/`temp_hidden_states` and `freq_attentions`/`temp_attentions`).

Do either of you have strong opinions on maintaining the same signature for the outputs as the rest of the lib? Or happy to go with separate frequency/temporal outputs?

Reply:
I don't have any strong opinion on this as well, but I believe we should stick to the Transformers design ever as much as possible.

One way to do it, would be to stick to the classic [BaseModelOutput](https://huggingface.co/docs/transformers/v4.36.1/en/main_classes/output#transformers.modeling_outputs.BaseModelOutput) as much as possible, especially in the order we display the output: first `temp_last_hidden_states`, then `temp_hidden_states` then `temp_attentions` if any, and then the same thing for the frequency. In other words, we would chose one of the branch as the main one and the other as secondary outputs.

One other way would be to designe one temp_model_output and one freq_model_output that both derives from [BaseModelOutput](https://huggingface.co/docs/transformers/v4.36.1/en/main_classes/output#transformers.modeling_outputs.BaseModelOutput) and return both in another class ?

As said, I don't have any strong opinion, the model will depart from the Transformers design anyway, so it's only a matter of choosing one design that makes sense to us and to core maintainers and sticking to it.

Reply:
No strong opinions, just think that it makes sense to concat them as much as just processing both by calling the transformer twice!